sympy/concrete/products.py

```diff
@@ -399,6 +399,13 @@ def _eval_product_direct(self, term, limits):
         (k, a, n) = limits
         return Mul(*[term.subs(k, a + i) for i in range(n - a + 1)])
 
+    def _eval_derivative(self, x):
+        from sympy import Sum
+        limits = list(self.limits)
+        if any(x in limit[1:].free_symbols for limit in reversed(limits)):
+            return None
+        return self * Sum(self.function.diff(x) / self.function, self._args[1:], **self.assumptions0)
```

Comment:
Is there a way to rewrite this to avoid a possible division by 0? I think Piecewise might bee too cumbersome but I can't think of any other way to do it. Any suggestions?

Reply:
The `self.function == 0` case already gives answer zero as in ` diff(Product(0, (y, 1, z)), x)`:
```
In [1]: diff(Product(0, (y, 1, z)), x)
Out[1]: 0
```
I think `diff` handles it as I couldn't find anything in `Product` to return `0`.
Kindly share your opinion.
Thanks,

Reply:
I was thinking something like `diff(Product(i/x, (i, -1, 1)), x)` or something like that. But I guess it can never be explicitly zero. How you have it now should be fine then.

Reply:
The PR gives:
```python
In [5]: Product(sin(n*x), (n, -1, 1)).diff(x).doit()
Out[5]: nan

In [6]: Product(sin(n*x), (n, -1, 1)).doit().diff(x)
Out[6]: 0

In [7]: Product(sin(n*x), (n, -1, 1)).diff(x)
Out[7]: 
                    1              
                   ____            
                   ╲               
⎛  1            ⎞   ╲              
⎜─┬─┬─          ⎟    ╲   n⋅cos(n⋅x)
⎜ │ │   sin(n⋅x)⎟⋅   ╱   ──────────
⎜ │ │           ⎟   ╱     sin(n⋅x) 
⎝n = -1         ⎠  ╱               
                   ‾‾‾‾            
                  n = -1  
```
A slightly more complicated formula can be given that does not involve division:
https://en.wikipedia.org/wiki/Product_rule#Product_of_more_than_two_factors

Division would be safe if all `f_i(x)` are known to be nonzero for all `x`. You could check `self.is_zero is False`.

Reply:
Is there a way to tell SymPy to sum over `f(n)` except for a specific index?

Reply:
Not that I know of but it can also be two sums:
```python
In [31]: Sum(f(n), (n, 1, m-1)) + Sum(f(n), (n, m+1, k))
Out[31]: 
m - 1            k         
 ___            ___        
 ╲              ╲          
  ╲              ╲         
  ╱   f(n) +     ╱     f(n)
 ╱              ╱          
 ‾‾‾            ‾‾‾        
n = 1        n = m + 1     

In [32]: _.subs(m, 3).subs(k, 10).doit()
Out[32]: f(1) + f(2) + f(4) + f(5) + f(6) + f(7) + f(8) + f(9) + f(10)
```

Reply:
@oscarbenjamin, @Maelstrom6 I tried to implement https://en.wikipedia.org/wiki/General_Leibniz_rule#More_than_two_factors
```
def _eval_derivative(self, x):
        if self.is_zero:
            return S.Zero
        from sympy.concrete.summations import Sum, Dummy
        i = Dummy('i')
        f = self.function
        limits = list(self.limits)
        _, a, b = limits[0]
        if any(x in limit[1:].free_symbols for limit in limits):
            return None
        return Sum(f.subs(_, i).diff(x) * Product(f, (_, a, i-1)) * Product(f, (_, i+1, b)), (i, a, b),  **self.assumptions0)
```

This works for previous cases and also for `Product(sin(n*x), (n, -1, 1)).diff(x).doit()` as:
```
In [1]: Product(sin(n*x), (n, -1, 1)).diff(x)
Out[1]:
  1
_____
╲
 ╲
  ╲               ⎛i - 1          ⎞     1
   ╲              ⎜─┬─┬─          ⎟   ─┬─┬─
   ╱   i⋅cos(i⋅x)⋅⎜ │ │   sin(n⋅x)⎟⋅   │ │    sin(n⋅x)
  ╱               ⎜ │ │           ⎟    │ │
 ╱                ⎝n = -1         ⎠ n = i + 1
╱
‾‾‾‾‾
i = -1

In [2]: Product(sin(n*x), (n, -1, 1)).diff(x).doit()
Out[2]: 0
```